crates/wasmi/src/module/instantiate/mod.rs

```diff
@@ -204,17 +212,16 @@ impl Module {
         &self,
         context: &mut impl AsContextMut,
         builder: &mut InstanceEntityBuilder,
-    ) {
+    ) -> Result<(), MemoryError> {
+        context
+            .as_context_mut()
+            .store
+            .check_new_memories_limit(self.len_memories())?;
         for memory_type in self.internal_memories().copied() {
-            let memory =
-                Memory::new(context.as_context_mut(), memory_type).unwrap_or_else(|error| {
-                    panic!(
-                        "encountered unexpected invalid memory type \
-                        {memory_type:?} after Wasm validation: {error}",
-                    )
-                });
+            let memory = Memory::new(context.as_context_mut(), memory_type)?;
```

Comment:
The `panic` was meaningful since it was a bug and not an error if this failed.

Reply:
Hah, this line is actually the entire point of the PR to me! Specifically being able to trap it nicely as an error.

I want to be able to catch (and prevent) instantiating a `Memory` that is too big according to the `ResourceLimiter`. Inside `Memory::new` (actually in its subsequent callee, `MemoryEntity::new`) the `ResourceLimiter` will return an error if it does not like the size of the memory, by policy. That's not a bug, that's a runtime limit-enforcement.

I understand that it subsumes the case that you were catching before -- if I'm reading correctly, the only bug it caught before was if you had mis-coded the validator and let through a page count that, translated to bytes, exceeds 4GiB  -- so I _think_ the thing for me to do here, if you want to preserve that cross-check, is to sink the panic down into the line in `MemoryEntity::new` where the page count is converted to bytes.

The other option I can think of is to add a special `MemoryError` that _just_ catches that case, and feed it back out to the call site in `instantiate`. This seems a little odd but it might be more appropriate if you can think of cases where `MemoryEntity::new` _is_ called (even indirectly) with a `MemoryType` that was not supposed to have already been validated.

I'm happy to go either way. Let me know!

Reply:
Ah, I should not review PRs as late as 1AM. Forgive me, you are obviously totally right. The change is as intended by the PR.